README.rst
======
README
======

boktaisim is a small tkinter application and library to make playing the Boktai trilogy of games
more easy. For more information, read through the `Background` section.

|
|

-----------
Screenshots
-----------

Boktai 1 using manual data

|boktai1|

Boktai 2 using zipcode

|boktai2|

Options screen

|options|

|
|

----------
Background
----------

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Nathan Stiles, Ideas and Design
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The initial goal of this project was to create a way for me to play and record the Boktai trilogy that was reminiscent of the original intent, without the need for sunlight. While I could never properly preserve the true Boktai experience of using a flaky solar sensor, contending with the glare of the sun on the GBA screen, and slowly succumbing to sun burn... I could at least come up with an alternative. Of course I couldn't make an actual solar sensor to be used in doors, so I chose to base my 'sun reading' on weather instead. I later found out that Boktai 4, known as Lunar Knights outside of Japan, handled it's energy meters in a similar way, which was a happy accident.

The more I thought about it, the more I realized how useful this tool could be to Boktai fans around the world. Boktai and Boktai 2 carts are becoming harder and harder to come by, and Boktai 3 never left Japan, requiring a fan translation to be enjoyed by anyone who can't read the language. To make matters worse, the carts that are available won't last forever: their batteries are dying and I don't how long the life of the solar sensors themselves are. So I've decided to make this alternative, imperfect as it may be, public for anyone to use!

The first and most important thing to note about the program is that it does not interact with the rom or emulator in any way. I have been playing the game through Retroarch_ using the mGBA_ core which has a sunlight sensor setting. You can actually map the raising and lowering of the sun meter to buttons on your keyboard or controller to fill the bar manually. Personally, I didn't feel it was worth the extra effort to code a connection through Bizhawk_ when a couple of simple button presses achieves the same solution with an emulator that I already know has the features I want and need. Obviously this program is open source though, so if you want to take a crack at it, then by all means do so!

What the app does do is generate an artificial sunlight reading that you can manually input into the game as you play. The program will give a chime at set intervals if there is a change to how much sunlight your in-game meter should have. This allows you to keep it open in the background while you play and not have to worry about checking it manually every couple of minutes. To generate the readings we used a formula combining temperature relative to the low and high forecast for the day, weather status (Clear Skies, Partial Clouds, Rain etc), position of the sun from sunrise to sunset, and a bit of RNG to determine how much 'sunlight' the meter is getting at any one time. We thought that these particular parameters would make the game playable by everyone around the world, as opposed to just naturally 'hot' environments with clear skies all the time (though, those areas will tend to have higher average readings of course).

Some of you may be asking why I'd go through all this trouble when the `Myboy GBA emulator`_ actually has a setting that uses the phones light sensor to fill Boktai's sun meter based on how much light it sees. I debated playing this way, but I just can't get in the groove of playing an action adventure game with touch controls. It also doesn't fit with the theming of Boktai very well, in my opinion. The camera will use any light, not just UV rays, so it's not really based around nature anymore. This may be nitpicky, but it felt less authentic to me. That said if you are fine with playing on your smart phone, Myboy is a great alternative to this project, and arguably a simpler solution.

The only other thing you may be scratching your head about after looking at the program is... 'Lunar Mode'. Boktai, by default, is not meant to be played for long periods at night; In fact, some parts of the game are impossible to progress without sunlight. By default the program reflects this, but after further thinking we decided to implement an optional 'Lunar Mode' to make the game playable at all hours. We tried to keep the spirit of the game and program intact by making the lunar readings cap out at a much lower level than they would during the day, while still making the game fully playable, if a bit more challenging.

I think that just about covers it. If you are feeling generous you can check out my `Youtube channel`_ for some in-depth analysis of RPGs or more casual reviews of other content... like Boktai once those videos are done! Now if you'll excuse me, I've got more Boktai to play. Thank you all for reading!

-- Nathan Stiles, of Stiles' Reviews

|
|

~~~~~~~~~~~~~~~~~~~~~~
Connor Barthold, Coder
~~~~~~~~~~~~~~~~~~~~~~

It seemed like a neat project.

-- Connor Barthold, of The Internet

|
|

-----
Setup
-----

No special setup is required, everything should run out of the box after installation or download. If you want to view the distributions of possible values available, `matplotlib`_ is required.

|
|

-------------
Configuration
-------------

A default config file will be generated on first run and stored in the user's application data directory (utilizing the `appdirs`_ package). This file contains both configuration settings and the last input data, and is saved automatically any time either of these changes.

|
|

-----
Usage
-----

~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running as a GUI Application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is the main way the package is intended to be used. Simply run `boktaisim` after installing with `python setup.py install` to launch the tkinter interface. Alternatively, launch `boktaisim.app` or `boktaisim.exe` if using one of the pre-packaged executables.

|
|

~~~~~~~~~~~~~~~~~~~~~~
Importing as a Library
~~~~~~~~~~~~~~~~~~~~~~

The `BoktaiSim` class can be imported and used directly without a GUI interface.

>>> from boktaisim.classes import BoktaiSim
>>> sim = BoktaiSim(version=2, zipcode=90210)
>>> sim.value
6
>>>

|
|

----------------
Value Generation
----------------

The value generated by boktaisim is the weighted average of four values::

 temperature    The high, low, and current temperature value is taken from weather API data.
                These values are scaled to between 0 and 10 (or 0 to 8 for Boktai 1).
 weather        The high, low, and current temperature value is taken from weather API data.
                These values are then scaled to between the current weather state's min and max
                values taken from a static dict in `boktaisim.constants`.
 sun            The current position of the sun is used as an input to a function which picks a
                random value from a beta distribution that changes based on sun position.
                For more information, take a look at `BoktaiSim._calulate_sun_value`
 random         The min, max, and average for the current weather state are taken from a static
                dict in `boktaisim.constants`. These values are passed to `random.triangular`
                and the result is used.

|
|

---------
Packaging
---------

~~~~~~~
For Mac
~~~~~~~

::

  python setup.py py2app

|
|

~~~~~~~~~~~
For Windows
~~~~~~~~~~~

::

  python setup.py py2exe

|
|

-----
To-Do
-----

Pull requests are always welcome!

 - Update bar on version selection
 - Better code documentation
 - Make Fahrenheit/Celsius handling more sane
 - refactor tkinter code, it's a mess
 - standardize "lat/lon", "lat/long", etc.
 - rewrite `__str__` for `BoktaiSim`
 - Support for console mode

|
|

----------------
Acknowledgements
----------------

This program used to utilize the free weather API provided by https://www.metaweather.com/ (built by Jason_
from Potato_), however it seems to have disappeared. In it's place, it uses the Open-Meteo Weather Forecast
API provided by https://open-meteo.com/.

For translating lat/long to place names, the free Geocoding API by Maps.co is used: https://geocode.maps.co/.

Weather icons were taken from `Font Awesome`_, with some adaptations.

Because of difficulties with packaging on Mac, this program includes a copy of `zipcodes.db` from
the `pyzipcode project`_.


.. |boktai1| image:: screenshots/boktai1_manual.png
.. |boktai2| image:: screenshots/boktai2_zip.png
.. |options| image:: screenshots/options.png

.. _Retroarch: https://www.retroarch.com/
.. _mGBA: https://mgba.io/
.. _Bizhawk: http://tasvideos.org/BizHawk.html
.. _Myboy GBA emulator: https://play.google.com/store/apps/details?id=com.fastemulator.gbafree
.. _Youtube channel: https://www.youtube.com/c/stilesreviews
.. _matplotlib: https://matplotlib.org/
.. _appdirs: https://github.com/ActiveState/appdirs
.. _Jason: https://www.jasoncartwright.com/
.. _Potato: https://p.ota.to/
.. _Font Awesome: https://fontawesome.com/
.. _pyzipcode project: https://github.com/vangheem/pyzipcode


CHANGELOG.rst
=========
Changelog
=========

|
|

--------------------------
Version 0.1.2 (2022-07-18)
--------------------------

* Switched from metaweather API to open-meteo API


--------------------------
Version 0.1.1 (2021-07-27)
--------------------------

* First public release


--------------------------
Version 0.1.0 (2021-06-24)
--------------------------

* First commit



boktaisim/tests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import unittest

from .classes import BoktaiSim, WeatherInfo
from .constants import WEATHER_STATES

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider
from matplotlib.ticker import AutoMinorLocator


def comprehensive_test():
    fig, axs = plt.subplots(3, 3, figsize=(10, 10))
    plt.subplots_adjust(left=0.05, bottom=0.15, right=0.95, top=0.95, wspace=0.2, hspace=0.3)

    ax_sun = plt.axes([0.2, 0.06, 0.65, 0.01], facecolor='lightgoldenrodyellow')
    slider_sun = Slider(ax_sun, 'Sun Position', 0., 100.,
                        valinit=50, valstep=1)

    ax_max_temp = plt.axes([0.2, 0.047, 0.65, 0.01], facecolor='lightgoldenrodyellow')
    slider_max_temp = Slider(ax_max_temp, 'Max Temperature', 0., 100.,
                             valinit=90., valstep=None)

    ax_avg_temp = plt.axes([0.2, 0.034, 0.65, 0.01], facecolor='lightgoldenrodyellow')
    slider_avg_temp = Slider(ax_avg_temp, 'Avg Temperature', 0., 100.,
                             valinit=50., valstep=None)

    ax_min_temp = plt.axes([0.2, 0.021, 0.65, 0.01], facecolor='lightgoldenrodyellow')
    slider_min_temp = Slider(ax_min_temp, 'Min Temperature', 0., 100.,
                             valinit=0., valstep=None)

    ax_temp = plt.axes([0.2, 0.008, 0.65, 0.01], facecolor='lightgoldenrodyellow')
    slider_temp = Slider(ax_temp, 'Current Temperature', 0., 100.,
                         valinit=50., valstep=None)

    colors = {
        'temperature': 'red',
        'weather': 'blue',
        'sun_location': 'orange',
        'random': 'green',
        'total': 'gray'
    }
    labels = list(colors.keys())

    global min_temp, avg_temp, max_temp, current_temp, sun_value
    min_temp = 0.
    avg_temp = 50.
    max_temp = 100.
    current_temp = 50.
    sun_value = 50

    def _wrap_update(variable: str):
        def update(value=None):
            global min_temp, avg_temp, max_temp, current_temp, sun_value
            if variable == 'current_temp':
                current_temp = value
            if variable == 'min_temp':
                min_temp = value
            if variable == 'avg_temp':
                avg_temp = value
            if variable == 'max_temp':
                max_temp = value
            if variable == 'sun_value':
                sun_value = value
            i = 0
            j = 0
            for state in WEATHER_STATES:
                if state == 'sl':
                    continue
                results = general_test(
                    weather_state=state, min_temp=min_temp, max_temp=max_temp, avg_temp=avg_temp,
                    current_temp=current_temp, sun_value=sun_value, count=1000, version=1
                )
                axs[j, i].clear()
                axs[j, i].set_xticks(results[1])
                axs[j, i].xaxis.set_minor_locator(AutoMinorLocator(n=2))
                axs[j, i].tick_params(which='minor', direction='in', length=7, top=False)
                axs[j, i].tick_params(which='major', width=2, top=False, right=False)
                axs[j, i].bar(
                    results[1] - 3 / 8, results[0]['temperature'], align='center', width=1 / 4,
                    alpha=.75,
                    color=colors['temperature']
                )
                axs[j, i].bar(
                    results[1] - 1 / 8, results[0]['weather'], align='center', width=1 / 4,
                    alpha=.75,
                    color=colors['weather']
                )
                axs[j, i].bar(
                    results[1] + 1 / 8, results[0]['sun_location'], width=1 / 4, align='center',
                    alpha=.75, color=colors['sun_location']
                )
                axs[j, i].bar(
                    results[1] + 3 / 8, results[0]['random'], align='center', width=1 / 4,
                    alpha=.75,
                    color=colors['random']
                )
                axs[j, i].bar(
                    results[1], results[0]['total'], align='center', width=1, alpha=.45,
                    ls='dashed', color=colors['total'], edgecolor='black'
                )
                axs[j, i].set_title(WEATHER_STATES[state]['name'])
                i += 1
                if i >= 3:
                    j += 1
                    i = 0
            fig.canvas.draw_idle()

        return update

    slider_temp.on_changed(_wrap_update('current_temp'))
    slider_min_temp.on_changed(_wrap_update('min_value'))
    slider_avg_temp.on_changed(_wrap_update('avg_value'))
    slider_max_temp.on_changed(_wrap_update('max_value'))
    slider_sun.on_changed(_wrap_update('sun_value'))
    _wrap_update('current_temp')(current_temp)

    handles = [plt.Rectangle((50, 50), 5, 5, color=colors[label]) for label in labels]
    plt.legend(handles, labels, loc=(0, 8), mode='expand', ncol=5)
    plt.show()


def general_test(
        min_temp: float = 0,
        avg_temp: float = 20,
        max_temp: float = 35,
        current_temp: float = 12,
        weather_state: str = 'c',
        sun_value: float = 50,
        version: int = 2,
        count: int = 100,
        **kwargs
):
    sim = BoktaiSim(
        version=version,
        manual_data=WeatherInfo(
            state='N/A',
            city='N/A',
            latlong='manual',
            woeid='0',
            min_temp=min_temp,
            max_temp=max_temp,
            current_temp=current_temp,
            visibility=5,
            weather_state=weather_state,
            sunrise='2021-06-20T04:19:57.380989-08:00',
            sunset='2021-06-20T23:42:08.855441-08:00',
            timestamp='2021-06-20T22:32:22.441253-08:00',
            manual=True,
            avg_temp=avg_temp
        )
    )

    i = 0
    results = {
        'temperature': {},
        'weather': {},
        'sun_location': {},
        'random': {},
        'total': {}
    }
    for value_type in ('temperature', 'weather', 'sun_location', 'random', 'total'):
        for j in range(0, 11):
            results[value_type][j] = 0
    while i < count:
        values = {
            'temperature': sim.temperature_value,
            'weather': sim.weather_value,
            'sun_location': sim._calulate_sun_value(sun_value, **kwargs),
            'random': sim.random_weather_value
        }
        # total = (values['temperature'] * 10) + (values['weather'] * 20) + (values['random'] * 25) + (values[
            # 'sun_location'] * 50)
        total = (values['temperature'] * 15) + (values['weather'] * 25) + (values['random'] * 20) + (values[
            'sun_location'] * 40)
        result = total / 100
        if weather_state in ('h', 't'):
            result -= 2
            if result < 0:
                result = 0
        if weather_state in ('hr', 'hc'):
            result -= 1
            if result < 0:
                result = 0
        if weather_state in ('lc', 's', 'c'):
            result += 1
            if result > 10:
                result = 10
        values['total'] = result
        for value_type in ('temperature', 'weather', 'sun_location', 'random', 'total'):
            results[value_type][round(values[value_type])] += 1
        i += 1
    y_pos = np.arange(11)

    dist = {}
    for value_type in ('temperature', 'weather', 'sun_location', 'random', 'total'):
        dist[value_type] = []
        for i in range(0, 11):
            if i in results[value_type]:
                dist[value_type].append(results[value_type][i])
            else:
                dist[value_type].append(0)

    return dist, y_pos


def sun_curve_test(sun_value: float = 50, count: int = 100, **kwargs):
    sim = BoktaiSim(
        version=2,
        manual_data=WeatherInfo(
            state='N/A',
            city='N/A',
            latlong='manual',
            woeid='0',
            min_temp=0,
            max_temp=35,
            current_temp=12,
            visibility=5,
            weather_state='c',
            sunrise='2021-06-20T04:19:57.380989-08:00',
            sunset='2021-06-20T23:42:08.855441-08:00',
            timestamp='2021-06-20T22:32:22.441253-08:00',
            manual=True,
            avg_temp=20
        )
    )

    i = 0
    results = {}
    while i < count:
        result = round(sim._calulate_sun_value(sun_value, **kwargs))
        if result not in results:
            results[result] = 1
        else:
            results[result] += 1
        i += 1
    dist = []
    for i in range(0, 11):
        if i in results:
            dist.append(results[i])
        else:
            dist.append(0)

    objects = tuple(range(0, 11))
    y_pos = np.arange(len(objects))
    return dist, y_pos


def location_value_test(zipcode: int, count: int = 100, lunar_mode: bool = False, **kwargs):
    sim = BoktaiSim(version=2, zipcode=zipcode, lunar_mode=lunar_mode)
    i = 0
    results = {}
    while i < count:
        result = sim.value
        if result not in results:
            results[result] = 1
        else:
            results[result] += 1
        i += 1
    print(results)
    dist = []
    for i in range(0, 11):
        if i in results:
            dist.append(results[i])
        else:
            dist.append(0)
    print(dist)

    objects = tuple(range(0, 11))
    y_pos = np.arange(len(objects))

    plt.bar(y_pos, dist, align='center', alpha=0.5)
    plt.xticks(y_pos, tuple(range(0, 11)))
    plt.ylabel('Occurrence')
    plt.title(f'BoktaiSim Value distribution\nLocation: {zipcode}, Lunar: {lunar_mode}\n'
              f'Weather: {WEATHER_STATES[sim.weather.weather_state]["name"]}\n'
              f'Min: {sim.weather.min_temp_f}, Current: {sim.weather.current_temp_f}, Max: '
              f'{sim.weather.max_temp_f}')

    plt.show()


boktaisim/constants.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import datetime

LOCAL_TIMEZONE = datetime.datetime.now(datetime.timezone.utc).astimezone().tzinfo

# Openmeteo uses different weather codes than metaweather, and they don't always neatly mesh.
OPENMETEO_WEATHER_STATES = {
    0: 'c',     # Clear Sky
    1: 'lc',    # Mainly Clear
    2: 'lc',    # Partly Cloudy
    3: 'hc',    # Overcast
    45: 'hc',   # Fog
    48: 'hc',   # Depositing Rime Fog (Freezing Ice Fog)
    51: 's',    # Light Drizzle
    53: 's',    # Moderate Drizzle
    55: 'lr',   # Dense Drizzle
    56: 's',    # Light Freezing Drizzle
    57: 'lr',   # Dense Freezing Drizzle
    61: 'lr',   # Slight Rain
    63: 'hr',   # Moderate Rain
    65: 'hr',   # Heavy Rain
    66: 'lr',   # Light Freezing Rain
    67: 'h',    # Heavy Freezing Rain
    71: 'sn',   # Slight Snow Fall
    73: 'sn',   # Moderate Snow Fall
    75: 'sn',   # heavy Snow Fall
    77: 'sn',   # Snow Grains
    80: 'hr',   # Slight Rain Showers
    81: 'hr',   # Moderate Rain Showers
    82: 'hr',   # Violent Rain Showers
    85: 'sl',   # Slight Snow Showers
    86: 'h',    # Heavy Snow Showers
    95: 't',    # Slight or Moderate Thunderstorm
    96: 't',    # Thunderstorm with Slight Hail
    99: 't'     # Thunderstorm with Heavy Hail
}

WEATHER_STATES = {
    'sn': {
        'name': 'Snow',
        'min': 0,
        'avg': 2,
        'max': 5,
        'mod': 0
    },
    'sl': {
        'name': 'Sleet',
        'min': 0,
        'avg': 1,
        'max': 3,
        'mod': -2
    },
    'h': {
        'name': 'Hail',
        'min': 0,
        'avg': 1,
        'max': 3,
        'mod': -2
    },
    't': {
        'name': 'Thunderstorm',
        'min': 0,
        'avg': 1,
        'max': 2,
        'mod': -2
    },
    'hr': {
        'name': 'Heavy Rain',
        'min': 0,
        'avg': 2,
        'max': 4,
        'mod': -1
    },
    'lr': {
        'name': 'Light Rain',
        'min': 0,
        'avg': 3,
        'max': 8,
        'mod': 0
    },
    's': {
        'name': 'Showers',
        'min': 1,
        'avg': 5,
        'max': 9,
        'mod': 1
    },
    'hc': {
        'name': 'Heavy Cloud',
        'min': 0,
        'avg': 2,
        'max': 4,
        'mod': -1
    },
    'lc': {
        'name': 'Light Cloud',
        'min': 2,
        'avg': 5,
        'max': 10,
        'mod': 1
    },
    'c': {
        'name': 'Clear',
        'min': 4,
        'avg': 7,
        'max': 10,
        'mod': 1
    },
}

WEATHER_STATES_REVERSE = {
    'Snow': 'sn',
    'Sleet': 'sl',
    'Hail': 'h',
    'Thunderstorm': 't',
    'Heavy Rain': 'hr',
    'Light Rain': 'lr',
    'Showers': 's',
    'Heavy Cloud': 'hc',
    'Light Cloud': 'lc',
    'Clear': 'c'
}

SUN_STATES = {
    'sunrise': {
        'name': 'Sunrise',
        'icon': 'Rising.gif'
    },
    'rise': {
        'name': 'Rising',
        'icon': 'Rising.gif'
    },
    'apex': {
        'name': 'At Apex',
        'icon': 'At Apex.gif'
    },
    'descend': {
        'name': 'Descending',
        'icon': 'Descending.gif'
    },
    'sunset': {
        'name': 'Sunset',
        'icon': 'Descending.gif'
    },
    'moon': {
        'name': 'Moonlight',
        'icon': 'Moonlight.gif'
    },
}

FEATURE_WEIGHTS = {
    'temperature': 10,
    'weather': 20,
    'sun_location': 40,
    'random': 25
}

# Pixel offsets for the stages of the various meters
BOKTAI_METER = {
    1: {
        0: 0,
        1: 77,
        2: 104,
        3: 131,
        4: 158,
        5: 185,
        6: 212,
        7: 239,
        8: 266
    },
    2: {
        0: 0,
        1: 30,
        2: 57,
        3: 84,
        4: 111,
        5: 138,
        6: 165,
        7: 192,
        8: 219,
        9: 246,
        10: 275
    },
    3: {
        0: 0,
        1: 30,
        2: 57,
        3: 84,
        4: 111,
        5: 138,
        6: 165,
        7: 192,
        8: 219,
        9: 246,
        10: 275
    }
}

# List of all images to load at start
IMAGES = [
    'boktai1_logo.gif',
    'boktai2_logo.gif',
    'boktai3_logo.gif',
    'boktai1_meter_empty.jpg',
    'boktai1_meter_full.jpg',
    'boktai2_meter_empty.jpg',
    'boktai2_meter_full.jpg',
    'boktai3_meter_empty.jpg',
    'boktai3_meter_full.jpg',
    'Solar_Sensor_Icon.gif',
    'boktaisim_icon.gif',
    'boktaisim_icon.ico',
    'boktaisim_icon.icns',
    'Rising.gif',
    'At Apex.gif',
    'Descending.gif',
    'Moonlight.gif',
    'sn.gif',
    'sl.gif',
    'h.gif',
    't.gif',
    'hr.gif',
    'lr.gif',
    's.gif',
    'hc.gif',
    'lc.gif',
    'c.gif'
]

SOUNDS = {
    'open': {
        'file': 'open.wav',
        'type': 'flavor'
    },
    'bar_update': {
        'file': 'chime1.wav',
        'type': 'alert'
    },
    'warning': {
        'file': 'overheat.wav',
        'type': 'alert'
    },
    'about': {
        'file': 'otenko.wav',
        'type': 'flavor'
    },
    'close': {
        'file': 'close.wav',
        'type': 'flavor'
    }
}
